File: Year 2/Semester 4/MPP/SeventhWeekCSharpServerClient/networking/utils/AbstractServer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Net;
using System.Text;
using System.Threading.Tasks;
using log4net.Repository.Hierarchy;
using System.IO;
using log4net;
using log4net.Util;

namespace SeventhWeekCSharpServerClient.networking.utils
{
    public abstract class AbstractServer
    {
        private static readonly ILog logger = LogManager.GetLogger("AbstractServer");
        private TcpListener server;
        private string host;
        private int port;
        public AbstractServer(string host, int port)
        {
            this.host = host;
            this.port = port;
        }
        public void Start()
        {
            IPAddress adr = IPAddress.Parse(host);
            IPEndPoint ep = new IPEndPoint(adr, port);
            try
            {
                server = new TcpListener(ep);
                server.Start();
                while (true)
                {
                    logger.Info("Waiting for clients ...");
                    TcpClient client = server.AcceptTcpClient();
                    logger.Info("Client connected ...");
                    ProcessRequest(client);
                }
            }
            catch (IOException e)
            {
                logger.Error("Error starting server", e);
                throw new ServerException("Starting server error ", e);
            }
        }

        public abstract void ProcessRequest(TcpClient client);

        public void Stop()
        {
            try
            {
                server.Stop();
                logger.Info("Server stopped");
            }
            catch (IOException e)
            {
                logger.Error("Error stopping server", e);
                throw new ServerException("Closing server error ", e);
            }
        }
    }
}

File: Year 2/Semester 4/Programming and Projection Environments/Exam preparation/Generare numere random/Martin - 2023 - pozitii/Networking/src/main/java/martin/network/utils/AbstractServer.java
package martin.network.utils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;


public abstract class AbstractServer {
    private final int port;
    private ServerSocket server=null;
    private static final Logger logger= LogManager.getLogger();
    public AbstractServer( int port){
              this.port=port;
    }

    public void start() throws ServerException {
        try{
            server=new ServerSocket(port);
            while(true){
                logger.info("Waiting for clients ...");
                Socket client=server.accept();
                logger.info("Client connected ...");
                processRequest(client);
            }
        } catch (IOException e) {
            logger.error("Error starting server",e);
            throw new ServerException("Starting server errror ",e);
        }finally {
            stop();
        }
    }

    protected abstract void processRequest(Socket client);
    public void stop() throws ServerException {
        try {
            server.close();
            logger.info("Server stopped");
        } catch (IOException e) {
            logger.error("Error stopping server",e);
            throw new ServerException("Closing server error ", e);
        }
    }
}
